README
all documentation moved to doc folder

ChangeLog
#    SQUID Acconting                                            [SAcc system]
#   Copyright (C) 2003-2006  the.Virus
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
version notes:
 STABLE - production releases was tested with default values. (on OpenBSD and/or FreeBSD and/or Linux),
 PRE - prerelease, partially tested (I think much more then CURRENT)
 CURRENT - may have some bugs, may not works. but "make" is passed.
 CVS - full UNTESTED version.

Versions changing.
// www: http://sacc.cybersec.ru
//$Author: slavik $ $Date: 2008-01-16 14:25:35 +0500 (п║я─п╢, 16 п╞п╫п╡ 2008) $
//$Id: ChangeLog 17 2008-01-16 09:25:35Z slavik $
// (C) the.Virus
// Legend:
// + Added
// * Improved/changed
// - Bug fixed
// ! Important

20-08-2008 1.80-STABLE3
- Исправлена ошибка с обработкой логов большого размера
+ Импорт и экспорт пользователей (для обновления системы)
- Исправлена ошибка с ротированием лога в конце месяца
* Улучшен автопоиск файлов сквида для скрипта configure
! Отныне цель upgrade работает только в рамках одного релиза (т.е. например с 1.80-STABLE1 до 1.80-STABLE3)

30-03-2007 1.80STABLE2
- Исправлена ошибка с обработкой спецсимволов в именах пользователей

16-03-2006 1.80stable1
- Исправлена ошибка при которой по умолчанию был режим счета по IP

18-10-2006 1.80rc3
* добавлен вывод отладочной инфы о поиске mysql.
* добавлена возможность считать траффик без авторизации и без отключения.

26-05-2006 1.80rc2
- Добавлена возможность скрыть rsh stat если её нет
- Добавлен хэдер устанавливающий принудительно кодировку кои8
- Исправлена ошибка из-за которой некорректно передавался пароль на root от СУБД.
- Исправлена ошибка из-за которой писалось в syslog, хотя параметры явно указывали не делать этого.
- Немного изменены параметры расчета времени проверок, должно меньше стать накладных расходов
* Убрана необходимость включать RegisterGlobals в пхп.

28-03-2006 1.80rc1
- исправлены проблемы из-за которых не собиралось под gcc 4.1
+ создан механизм автопоиска mysql (уж очень заебали)
+ создан механизм автопоиска куда установлены apache и squid (за компанию с mysql)
- исправлены ворнинги компилятора при сборке на линуксе FC5, для красоты.
+ добавлен ключик который позволяет по существующему логу создать всех юзеров с дефолтными настройками, и в последствии создавать новых автоматически.
+ теперь пхп подхватывает registerglobals из .htaccess, правда для этого надо включить "AllowOverride All" в настройках апача

24-03-2006 1.80pre4
- добавлена совместимость с solaris
- исправлены потенциально несовместимые с другими платформами места кода
- исправлена логика ухода в daemon mode
- исправлено падение при отсутсвии в записей в таблице uhistory
- исправлено обновление БД с версии 1.704

18-01-2006 1.80pre3
+ добавлен вывод статистики по дням
- при чистой установке происходил сбой

5-11-2005 1.80pre2
! изменен формат таблиц.
* создан кэш идентификаторов пользователей, уменьшает нагрузку на субд.
+ добавлена возможность редактирования списков доступа.
+ добавлена история за прошлые месяцы.
+ добавлена статистика эффективности кэша за этот месяц.
+ теперь возможна настройка системы через веб.
+ теперь возможно управлять демоном через веб, а не только сигналами.

05-05-2005 1.80pre1
- compilable under mandrake 10
- web interface, NCSA mode, now administrator can reset user password
- web interface, NCSA mode, now administrator can't change user login
* can use remote mysql server, see configure --help for details

08-12-2004 1.72-stable4
- компилируется под freebsd 5.3
- убрано ограничение на пользователей с "-" в имени. (возможно будут записи о пользователе - в логе)
- теперь все аутентификаторы устанавливаются по умолчанию, ключи configure не играют роли, оставлены для настройки вебконсоли.

10-11-2004 1.72-stable3
- исправлены ошибки в обработке строк.
! обработка строк немного оптимизирована.
- исправлена проблема с неправильным кэшированием чтения на некоторых системах.
- исправлена проблема в конфигурационном скрипте, в подсказке было написано "--enable-mysql-prefix" реально ожидалось "--with-mysql-prefix"
! изменены ключи отладки, смотри --help для подробностей.
+ немного расширена документация и FAQ.
- исправлен косяк с неправильным отображением ip на детальной статистике, правда для этого нужно не просто апгрейдить исполняемый файл, а базу пересоздать...

20-08-2004 1.72-stable2
- Исправлено (в двух местах) вылетание при некорректных строках.
- Исправлена ошибка с необработкой ошибки EINTR, из-за которой слетал оффсет в некоторых случаях.
- Исправлена ошибка в инсталляционном скрипте.

2004-06-07 1.72-stable1
! initial release

28-05-2004 1.72-pre8
! doc's
- Setproctitle not used if not exist.
+ Added check for incorrect strings in ACCESS.LOG
* Now if pid of squid is incorrect - we silently skip managing.
* shorted application name for syslog.
* added reason for "incorrect string". (for testing)

13-04-2004 1.72-pre6
* better daemonization and fault control
- fixed broken web interface in pre5-p1
- error "<varargs.h> is obsolete with this version of GCC."
- some fixes in web interface
* ncsa - added password changing by SA (not tested)
! freezed developing sendmail maillog processing.

17-02-2004 1.72-pre5

src/auth/lib/md5.c
/*
 * $Id: md5.c 17 2008-01-16 09:25:35Z slavik $
 */

/* taken from RFC-1321/Appendix A.3 */

/*
 * MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm
 */

/*
 * Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All rights
 * reserved.
 * 
 * License to copy and use this software is granted provided that it is
 * identified as the "RSA Data Security, Inc. MD5 Message-Digest Algorithm"
 * in all material mentioning or referencing this software or this function.
 * 
 * License is also granted to make and use derivative works provided that such
 * works are identified as "derived from the RSA Data Security, Inc. MD5
 * Message-Digest Algorithm" in all material mentioning or referencing the
 * derived work.
 * 
 * RSA Data Security, Inc. makes no representations concerning either the
 * merchantability of this software or the suitability of this software for
 * any particular purpose. It is provided "as is" without express or implied
 * warranty of any kind.
 * 
 * These notices must be retained in any copies of any part of this
 * documentation and/or software.
 */

#include "config.h"

#if HAVE_STRING_H
#include <string.h>
#endif

#include "md5.h"

/*
 * Constants for MD5Transform routine.
 */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static void MD5Transform(u_int32_t[4], const unsigned char[64]);
static void Encode(unsigned char *, u_int32_t *, unsigned int);
static void Decode(u_int32_t *, const unsigned char *, unsigned int);

#if HAVE_MEMCPY
#define MD5_memcpy(to,from,count) memcpy(to,from,count)
#else
static void MD5_memcpy(unsigned char *, unsigned char *, unsigned int);
#endif

#if HAVE_MEMSET
#define MD5_memset(buf,chr,count) memset(buf,chr,count)
#else
static void MD5_memset(char *, int, unsigned int);
#endif

static unsigned char PADDING[64] =
{
    0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/*
 * F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/*
 * ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/*
 * FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4. Rotation is
 * separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (u_int32_t)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

/*
 * MD5 initialization. Begins an MD5 operation, writing a new context.
 */
void
MD5Init(MD5_CTX * context)
{
    context->count[0] = context->count[1] = 0;
    /*
     * Load magic initialization constants.
     */
    context->state[0] = 0x67452301;
    context->state[1] = 0xefcdab89;
    context->state[2] = 0x98badcfe;
    context->state[3] = 0x10325476;
}

/*
 * MD5 block update operation. Continues an MD5 message-digest operation,
 * processing another message block, and updating the context.
 */
void
MD5Update(MD5_CTX * context, const void *input_p, unsigned long inputLen)
{
    unsigned int i, index, partLen;
    const unsigned char *input = input_p;

    /* Compute number of bytes mod 64 */
    index = (unsigned int) ((context->count[0] >> 3) & 0x3F);

    /* Update number of bits */
    if ((context->count[0] += ((u_int32_t) inputLen << 3))
	< ((u_int32_t) inputLen << 3))
	context->count[1]++;
    context->count[1] += ((u_int32_t) inputLen >> 29);

    partLen = 64 - index;

    /*
     * Transform as many times as possible.
     */
    if (inputLen >= partLen) {
	MD5_memcpy(&context->buffer[index], input, partLen);
	MD5Transform(context->state, context->buffer);

	for (i = partLen; i + 63 < inputLen; i += 64)
	    MD5Transform(context->state, &input[i]);

	index = 0;
    } else
	i = 0;

    /* Buffer remaining input */
    MD5_memcpy(&context->buffer[index], &input[i], inputLen - i);
}

/*
 * MD5 finalization. Ends an MD5 message-digest operation, writing the the
 * message digest and zeroizing the context.
 */
void
MD5Final(unsigned char *digest, MD5_CTX * context)
{
    unsigned char bits[8];
    unsigned int index, padLen;

    /* Save number of bits */
    Encode(bits, context->count, 8);

    /*
     * Pad out to 56 mod 64.
     */
    index = (unsigned int) ((context->count[0] >> 3) & 0x3f);
    padLen = (index < 56) ? (56 - index) : (120 - index);
    MD5Update(context, PADDING, padLen);

    /* Append length (before padding) */
    MD5Update(context, bits, 8);
    /* Store state in digest */
    Encode(digest, context->state, 16);

    /*
     * Zeroize sensitive information.
     */
    MD5_memset((char *) context, 0, sizeof(*context));
}

/*
 * MD5 basic transformation. Transforms state based on block.
 */
static void
MD5Transform(u_int32_t state[4], const unsigned char block[64])
{
    u_int32_t a = state[0], b = state[1], c = state[2], d = state[3], x[16];

    Decode(x, block, 64);

    /* Round 1 */
    FF(a, b, c, d, x[0], S11, 0xd76aa478);	/* 1 */
    FF(d, a, b, c, x[1], S12, 0xe8c7b756);	/* 2 */
    FF(c, d, a, b, x[2], S13, 0x242070db);	/* 3 */
    FF(b, c, d, a, x[3], S14, 0xc1bdceee);	/* 4 */
    FF(a, b, c, d, x[4], S11, 0xf57c0faf);	/* 5 */
    FF(d, a, b, c, x[5], S12, 0x4787c62a);	/* 6 */
    FF(c, d, a, b, x[6], S13, 0xa8304613);	/* 7 */
    FF(b, c, d, a, x[7], S14, 0xfd469501);	/* 8 */
    FF(a, b, c, d, x[8], S11, 0x698098d8);	/* 9 */
    FF(d, a, b, c, x[9], S12, 0x8b44f7af);	/* 10 */
    FF(c, d, a, b, x[10], S13, 0xffff5bb1);	/* 11 */
    FF(b, c, d, a, x[11], S14, 0x895cd7be);	/* 12 */
    FF(a, b, c, d, x[12], S11, 0x6b901122);	/* 13 */
    FF(d, a, b, c, x[13], S12, 0xfd987193);	/* 14 */
    FF(c, d, a, b, x[14], S13, 0xa679438e);	/* 15 */
    FF(b, c, d, a, x[15], S14, 0x49b40821);	/* 16 */

    /* Round 2 */
    GG(a, b, c, d, x[1], S21, 0xf61e2562);	/* 17 */
    GG(d, a, b, c, x[6], S22, 0xc040b340);	/* 18 */
    GG(c, d, a, b, x[11], S23, 0x265e5a51);	/* 19 */
    GG(b, c, d, a, x[0], S24, 0xe9b6c7aa);	/* 20 */
    GG(a, b, c, d, x[5], S21, 0xd62f105d);	/* 21 */
    GG(d, a, b, c, x[10], S22, 0x2441453);	/* 22 */
    GG(c, d, a, b, x[15], S23, 0xd8a1e681);	/* 23 */
    GG(b, c, d, a, x[4], S24, 0xe7d3fbc8);	/* 24 */
    GG(a, b, c, d, x[9], S21, 0x21e1cde6);	/* 25 */
    GG(d, a, b, c, x[14], S22, 0xc33707d6);	/* 26 */
    GG(c, d, a, b, x[3], S23, 0xf4d50d87);	/* 27 */
    GG(b, c, d, a, x[8], S24, 0x455a14ed);	/* 28 */
    GG(a, b, c, d, x[13], S21, 0xa9e3e905);	/* 29 */
    GG(d, a, b, c, x[2], S22, 0xfcefa3f8);	/* 30 */
    GG(c, d, a, b, x[7], S23, 0x676f02d9);	/* 31 */
    GG(b, c, d, a, x[12], S24, 0x8d2a4c8a);	/* 32 */

    /* Round 3 */
    HH(a, b, c, d, x[5], S31, 0xfffa3942);	/* 33 */
    HH(d, a, b, c, x[8], S32, 0x8771f681);	/* 34 */
    HH(c, d, a, b, x[11], S33, 0x6d9d6122);	/* 35 */
    HH(b, c, d, a, x[14], S34, 0xfde5380c);	/* 36 */
    HH(a, b, c, d, x[1], S31, 0xa4beea44);	/* 37 */
    HH(d, a, b, c, x[4], S32, 0x4bdecfa9);	/* 38 */
    HH(c, d, a, b, x[7], S33, 0xf6bb4b60);	/* 39 */
    HH(b, c, d, a, x[10], S34, 0xbebfbc70);	/* 40 */
    HH(a, b, c, d, x[13], S31, 0x289b7ec6);	/* 41 */
    HH(d, a, b, c, x[0], S32, 0xeaa127fa);	/* 42 */
    HH(c, d, a, b, x[3], S33, 0xd4ef3085);	/* 43 */
    HH(b, c, d, a, x[6], S34, 0x4881d05);	/* 44 */
    HH(a, b, c, d, x[9], S31, 0xd9d4d039);	/* 45 */
    HH(d, a, b, c, x[12], S32, 0xe6db99e5);	/* 46 */
    HH(c, d, a, b, x[15], S33, 0x1fa27cf8);	/* 47 */
    HH(b, c, d, a, x[2], S34, 0xc4ac5665);	/* 48 */

    /* Round 4 */
    II(a, b, c, d, x[0], S41, 0xf4292244);	/* 49 */
    II(d, a, b, c, x[7], S42, 0x432aff97);	/* 50 */
    II(c, d, a, b, x[14], S43, 0xab9423a7);	/* 51 */
    II(b, c, d, a, x[5], S44, 0xfc93a039);	/* 52 */
    II(a, b, c, d, x[12], S41, 0x655b59c3);	/* 53 */
    II(d, a, b, c, x[3], S42, 0x8f0ccc92);	/* 54 */
    II(c, d, a, b, x[10], S43, 0xffeff47d);	/* 55 */
    II(b, c, d, a, x[1], S44, 0x85845dd1);	/* 56 */
    II(a, b, c, d, x[8], S41, 0x6fa87e4f);	/* 57 */
    II(d, a, b, c, x[15], S42, 0xfe2ce6e0);	/* 58 */
    II(c, d, a, b, x[6], S43, 0xa3014314);	/* 59 */
    II(b, c, d, a, x[13], S44, 0x4e0811a1);	/* 60 */
    II(a, b, c, d, x[4], S41, 0xf7537e82);	/* 61 */
    II(d, a, b, c, x[11], S42, 0xbd3af235);	/* 62 */
    II(c, d, a, b, x[2], S43, 0x2ad7d2bb);	/* 63 */
    II(b, c, d, a, x[9], S44, 0xeb86d391);	/* 64 */

    state[0] += a;
    state[1] += b;
    state[2] += c;
    state[3] += d;

    /*
     * Zeroize sensitive information.
     */
    MD5_memset((char *) x, 0, sizeof(x));
}

/*
 * Encodes input (u_int32_t) into output (unsigned char). Assumes len is a
 * multiple of 4.
 */
static void
Encode(unsigned char *output, u_int32_t * input, unsigned int len)
{
    unsigned int i, j;

    for (i = 0, j = 0; j < len; i++, j += 4) {
	output[j] = (unsigned char) (input[i] & 0xff);
	output[j + 1] = (unsigned char) ((input[i] >> 8) & 0xff);
	output[j + 2] = (unsigned char) ((input[i] >> 16) & 0xff);
	output[j + 3] = (unsigned char) ((input[i] >> 24) & 0xff);
    }
}

/*
 * Decodes input (unsigned char) into output (u_int32_t). Assumes len is a
 * multiple of 4.
 */
static void
Decode(u_int32_t * output, const unsigned char *input, unsigned int len)
{
    unsigned int i, j;

    for (i = 0, j = 0; j < len; i++, j += 4)
	output[i] = ((u_int32_t) input[j]) | (((u_int32_t) input[j + 1]) << 8) |
	    (((u_int32_t) input[j + 2]) << 16) | (((u_int32_t) input[j + 3]) << 24);
}

#if !HAVE_MEMCPY
/*
 * Note: Replace "for loop" with standard memcpy if possible.
 */

static void
MD5_memcpy(unsigned char *output, unsigned char *input, unsigned int len)
{
    unsigned int i;
    for (i = 0; i < len; i++)
	output[i] = input[i];
}
#endif

#if !HAVE_MEMSET
/*
 * Note: Replace "for loop" with standard memset if possible.
 */
static void
MD5_memset(char *output, int value, unsigned int len)
{
    unsigned int i;
    for (i = 0; i < len; i++)
	output[i] = (char) value;
}

#endif
